model/task/task.go

```diff
@@ -2513,82 +2515,84 @@ func (t *Task) Insert() error {
 		if err != nil {
-			return errors.Wrap(err, "retrieving execution tasks")
-		}
-		if err = ArchiveMany(execTasks); err != nil {
-			return errors.Wrap(err, "archiving execution tasks")
+			grip.Error(message.WrapError(err, message.Fields{
+				"archive_task_id": archiveTask.Id,
+				"old_task_id":     archiveTask.OldTaskId,
+				"execution":       t.Execution,
+				"display_only":    t.DisplayOnly,
+			}))
+			return errors.Wrap(err, "inserting archived task into old tasks")
 		}
-	}
-
-	archiveTask := t.makeArchivedTask()
-	err := db.Insert(OldCollection, archiveTask)
-	if err != nil {
-		grip.Error(message.WrapError(err, message.Fields{
-			"archive_task_id": archiveTask.Id,
-			"old_task_id":     archiveTask.OldTaskId,
-			"execution":       t.Execution,
-			"display_only":    t.DisplayOnly,
-		}))
-		return errors.Wrap(err, "inserting archived task into old tasks")
-	}
-	err = UpdateOne(
-		bson.M{IdKey: t.Id},
-		bson.M{
-			"$unset": bson.M{
-				AbortedKey:              "",
-				AbortInfoKey:            "",
-				OverrideDependenciesKey: "",
-			},
-			"$inc": bson.M{ExecutionKey: 1},
-		})
-	if err != nil {
+		err = UpdateOne(bson.M{IdKey: t.Id},
+			bson.M{
+				"$unset": bson.M{
+					AbortedKey:              "",
+					AbortInfoKey:            "",
+					OverrideDependenciesKey: "",
+				},
+				"$inc": bson.M{ExecutionKey: 1},
+			})
+		t.Aborted = false
 		return errors.Wrap(err, "updating task")
 	}
-	t.Aborted = false
-	return nil
 }
 
+// ArchiveMany accepts tasks and display tasks (no execution tasks). The function
+// expects that each one is going to be archived and progressed to the next execution.
+// For execution tasks in display tasks, it will properly account for archiving
+// only tasks that should be if failed.
 func ArchiveMany(tasks []Task) error {
-	if len(tasks) == 0 {
-		return nil
-	}
-	existingTasksMap := map[string]bool{}
-	for _, t := range tasks {
-		existingTasksMap[t.Id] = true
-	}
-	additionalTasks := []string{}
+	allTaskIds := []string{}          // Contains all tasks and display tasks IDs
+	execTaskIds := []string{}         // Contains all exec tasks IDs
+	toUpdateExecTaskIds := []string{} // Contains all exec tasks IDs that should update and have new execution
+	archivedTasks := []interface{}{}  // Contains all archived tasks (task, display, and execution). Created by Task.makeArchivedTask()
+
 	for _, t := range tasks {
-		// For any display tasks here, make sure that we also archive all their execution tasks.
-		// Consider each execution tasks individually, so we don't query a task we already passed in.
-		for _, et := range t.ExecutionTasks {
-			if !existingTasksMap[et] {
-				additionalTasks = append(additionalTasks, et)
-				existingTasksMap[et] = true
+		allTaskIds = append(allTaskIds, t.Id)
+		archivedTasks = append(archivedTasks, t.makeArchivedTask())
+		if t.DisplayOnly && len(t.ExecutionTasks) > 0 {
+			var eTasks []Task
+			var err error
+
+			if t.ResetFailedWhenFinished {
```

Comment:
We should check ResetWhenFinished _before_ we check ResetFailedWhenFinished; if both are set, All would take precedence over just failed.